services/typeddata/hash_test.go

```diff
@@ -0,0 +1,261 @@
+package typeddata
+
+import (
+	"math/big"
+	"testing"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/stretchr/testify/require"
+)
+
+func TestTypeString(t *testing.T) {
+	type testCase struct {
+		description string
+		typeString  string
+		types       Types
+		target      string
+	}
+	for _, tc := range []testCase{
+		{
+			"WithoutDeps",
+			"Person(string name,address wallet)",
+			Types{"Person": []Field{{Name: "name", Type: "string"}, {Name: "wallet", Type: "address"}}},
+			"Person",
+		},
+		{
+			"SingleDep",
+			"Mail(Person from,Person to)Person(string name,address wallet)",
+			Types{
+				"Person": []Field{{Name: "name", Type: "string"}, {Name: "wallet", Type: "address"}},
+				"Mail":   []Field{{Name: "from", Type: "Person"}, {Name: "to", Type: "Person"}},
+			},
+			"Mail",
+		},
+		{
+			"DepsOrdered",
```

Comment:
What order do we test here which is not tested in `SingleDep`?

Reply:
but here i have two deps and they must be ordered, e.g. A then B. and in SingleDep there is only one

Reply:
`SingleDep` also tests the order as `Person` must be before `Mail`. But yeah, let's keep it, just wanted to make sure I understood it properly.

Reply:
in SingleDep Person is after Mail in the result. eip defines that primary type must be first, and all dependencies after that are sorted alphabetically.

Reply:
Oh I didn't know that it's alphabetically. I though it's evaluated based on the occurrence order. So `Z(B b, A a)A(string name)B(string name)` would still produce:
```
			Types{
				"A": []Field{{Name: "name", Type: "string"}},
				"B": []Field{{Name: "name", Type: "string"}},
				"Z": []Field{{Name: "a", Type: "A"}, {Name: "b", Type: "B"}},
			},
```
?

Reply:
yes, will add it to `deps` docstring